app/models/project.rb

```diff
@@ -93,14 +93,56 @@ def asset_count(session_id:)
     xml_metadata[:total_file_count]
   end
 
-  def file_list(session_id:, idx: 1, size: 10)
-    return {files: []} if mediaflux_id == nil
-    request = Mediaflux::Http::QueryRequest.new(
-      session_token: session_id,
-      collection: mediaflux_id,
-      idx:, size:,
-      action: "get-meta",
-      deep_search: true)
-    request.result
+  # Fetches the first n files
+  def file_list(session_id:, size: 10)
+    return { files: [] } if mediaflux_id.nil?
+
+    query_req = Mediaflux::Http::QueryRequest.new(session_token: session_id, collection: mediaflux_id, deep_search: true)
+    iterator_id = query_req.result
+
+    iterator_req = Mediaflux::Http::IteratorRequest.new(session_token: session_id, iterator: iterator_id, size: size)
+    results = iterator_req.result
+
+    # Destroy _after_ fetching the first set of results from iterator_req.
+    # This call is required since it possible that we have read less assets than
+    # what the collection has but we are done with the iterator.
+    Mediaflux::Http::IteratorDestroyRequest.new(session_token: session_id, iterator: iterator_id).resolve
+
+    results
   end
+
+  # Fetches the entire file list to a file
+  def file_list_to_file(session_id:, filename:)
+    return { files: [] } if mediaflux_id.nil?
+
+    query_req = Mediaflux::Http::QueryRequest.new(session_token: session_id, collection: mediaflux_id, deep_search: true)
+    iterator_id = query_req.result
+
+    File.open(filename, "w") do |file|
+      # file header
+      file.write("ID, PATH, NAME, COLLECTION?, LAST_MODIFIED, SIZE\r\n")
+      loop do
+        iterator_req = Mediaflux::Http::IteratorRequest.new(session_token: session_id, iterator: iterator_id, size: 1000)
```

Comment:
A more complete example, fetches the data from the iterator until the iterator says it has fetched all the data (completed)